src/background_workers.c

```diff
@@ -245,130 +385,148 @@ void *RedisAI_Run_ThreadMain(void *arg) {
+        if (dagRefCount == 0 && evicted_rinfo->client) {
+          RedisModule_UnblockClient(evicted_rinfo->client, evicted_rinfo);
         }
-        // If we're done processing a modelrun or scriptrun command
+      }
+
+      // If there was no progress on the DAG, meaning the inputs to the
+      // current operation were not ready (they depend on other workers on other
+      // queues completing their job), then we put the entry back on the queue
+      if (do_retry == 1) {
+        RedisAI_RunInfo *evicted_rinfo = (RedisAI_RunInfo *)(evicted_items[0]->value);
+
+        if (queueLength(run_queue_info->run_queue) > 0) {
+          // Pop the next item in the queue
+          queueItem *next_item = queuePop(run_queue_info->run_queue);
+          RedisAI_RunInfo *next_rinfo = (RedisAI_RunInfo *)next_item->value;
+          // Push the DAG to the front of the queue, and then the item we just
+          // popped in front of it, so that it becomes the first item in the queue.
+          // The rationale is, since the DAG needs to wait for other workers, we are
+          // giving way to the next item and we'll get back to the DAG when that is done
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+          queuePushFront(run_queue_info->run_queue, next_rinfo);
+        }
+        // If there's nothing else in the queue
         else {
-          // If the client is still connected
-          if (evicted_rinfo->client) {
-            // Unblock the client. Note that we do this for every evicted info in a batch,
-            // which will be associated with a different client. Here we unblock every
-            // client appropriately
-            RedisModule_UnblockClient(evicted_rinfo->client, evicted_rinfo);
+          // We push the DAG back at the front
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+          // Since there's nothing else on the queue we just break out and give other
+          // workers a chance to produce the inputs needed for this DAG step
+          array_free(evicted_items);
+          break;
+        }
+      }
+
+      // If the op was ran successfully and without any error, then put the entry back
+      // on the queue unless all ops for the device have been executed
+      if (do_run == 1 && run_error == 0) {
+        // Here we iterate backwards to keep the first evicted on top
+        // A side effect of this is that we are potentially changing priority in the queue
+        // We could solve this using a priority queue, TODO for later
+        for (long long i = array_len(evicted_items) - 1; i >= 0; i--) {
+          // Get the current evicted run info
+          RedisAI_RunInfo *evicted_rinfo = (RedisAI_RunInfo *)(evicted_items[i]->value);
+          // If the DAG for the top-most item is complete for the device, then
+          // we don't push it back on the queue
+          if (i ==0 && device_complete_after_run == 1) {
+            continue;
```

Comment:
why you re-enter an executed run info back in the queue?

Reply:
An individual rinfo represents a DAG on a device at a particular current op. When we run, we run the current op for the device. After that, we place the DAG back on the queue for the next round. Note that we need to do this because the next op might not be ready to run (waiting for other ops of the same DAG to execute on other devices), so the worker will have an opportunity to run another DAG in the meantime.